src/charm.py

```diff
@@ -99,86 +102,136 @@ def _configure_pebble(self) -> None:
             self._container.add_layer(self._container_name, layer, combine=True)
             self._container.restart(self._service_name)
 
-    def _configure_upf_information(self):
+    def _configure_upf_information(self) -> None:
         """The `fiveg_n4` relation is not mandatory.
 
-        If it exists it must contain the UPF hostname and port.
+        If it exists, config files are generated and pushed to the workload.
         """
-        if self.model.relations.get(FIVEG_N4_RELATION_NAME):
-            if not self._get_upf_hostname() or not self._get_upf_port():
-                logger.warning("Invalid information in %s integration", FIVEG_N4_RELATION_NAME)
+        if not self.model.relations.get(FIVEG_N4_RELATION_NAME):
+            logger.info("Relation %s not available", FIVEG_N4_RELATION_NAME)
+            return
+        upf_existing_content = self._get_existing_config_file(path=UPF_CONFIG_PATH)
+        upf_config_content = self._get_upf_config()
+        if not upf_config_content:
+            logger.error("UPF config file is invalid")
+            return
+        if not upf_existing_content or not config_file_content_matches(
+            existing_content=upf_existing_content, new_content=upf_config_content
+        ):
+            self._push_upf_config_file_to_workload(upf_config_content)
+
+    def _configure_gnb_information(self) -> None:
+        """The `fiveg_gnb_identity` relation is not mandatory.
 
-    def _get_upf_hostname(self) -> str:
-        """Gets UPF hostname from the `fiveg_n4` relation data bag.
+        If it exists, config files are generated and pushed to the workload.
+        """
+        if not self.model.relations.get(GNB_IDENTITY_RELATION_NAME):
+            logger.info("Relation %s not available", GNB_IDENTITY_RELATION_NAME)
+            return
+        gnb_existing_content = self._get_existing_config_file(path=GNB_CONFIG_PATH)
+        gnb_config_content = self._get_gnb_config()
+        if not gnb_config_content:
+            logger.error("gNB config file is invalid")
+            return
```

Comment:
If there is no gnb in the deployment, will we get into this condition? If so we should not log an error. There is a similar point to make in the upf.

Reply:
If there is not relation to gnb we will return earlier (line 128), if there is a gnb deployed we will try to create a config file by getting the gnb information from relation data, if the file is empty we log the error here, if the file is empty it will mean that we couldn't get info from the relation data.
So I think the design is correct, but I might be missing something? what do you think? @gruyaume